codegen/winterfell/src/air/graph.rs

```diff
@@ -147,3 +147,169 @@ fn binary_op_to_string(
         _ => panic!("unsupported operation"),
     }
 }
+
+// COMMON-SUBEXPRESSION ELIMINATION & CONVERSION HOISTING EMITTER
+// ================================================================================================
+// The functions below generate an expression string along with a list of temporary bindings which
+// should be emitted before using the expression. This reduces duplicate multiplications produced by
+// exponentiation lowering, and factors repeated extension field conversions (E::from(...)).
+
+use std::collections::HashMap;
+
+/// Returns a tuple of (temporary let bindings, final expression string) for the provided node.
+pub(super) fn to_string_with_temps(
+    ir: &Air,
+    node_index: &NodeIndex,
+    trace_segment: TraceSegmentId,
+) -> (Vec<String>, String) {
+    let mut use_counts: HashMap<NodeIndex, usize> = HashMap::new();
+    collect_use_counts(ir, node_index, &mut use_counts);
+
+    let mut cse_temps: HashMap<NodeIndex, String> = HashMap::new();
+    let mut conv_seen: HashMap<String, usize> = HashMap::new();
+    let mut conv_temps: HashMap<String, String> = HashMap::new();
+    let mut temps: Vec<String> = Vec::new();
+    let mut temp_counter: usize = 0;
+
+    let expr = emit_node(
+        ir,
+        node_index,
+        trace_segment,
+        &use_counts,
+        &mut cse_temps,
+        &mut conv_seen,
+        &mut conv_temps,
+        &mut temps,
+        &mut temp_counter,
+    );
+
+    (temps, expr)
+}
+
+fn collect_use_counts(
+    ir: &Air,
+    node_index: &NodeIndex,
+    use_counts: &mut HashMap<NodeIndex, usize>,
+) {
+    *use_counts.entry(*node_index).or_insert(0) += 1;
+    match ir.constraint_graph().node(node_index).op() {
+        Operation::Add(l, r) | Operation::Sub(l, r) | Operation::Mul(l, r) => {
+            collect_use_counts(ir, l, use_counts);
+            collect_use_counts(ir, r, use_counts);
+        },
+        Operation::Value(_) => {},
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn emit_node(
+    ir: &Air,
+    node_index: &NodeIndex,
+    trace_segment: TraceSegmentId,
+    use_counts: &HashMap<NodeIndex, usize>,
+    cse_temps: &mut HashMap<NodeIndex, String>,
```

Comment:
Testing should verify functional equivalence - the optimized generated code must produce identical results to the unoptimized version. 

You could implement similar verification by running both versions against the same inputs and comparing outputs.